### PR TITLE
OAuth2Manager Implementation

### PR DESCRIPTION
**Pull request by akanpatel2206**
API spec: #4772 
Issue: #441 

This new WinAppSDK version of the AuthManager is different than the existing public
[WebAuthenticationBroker](https://learn.microsoft.com/en-us/uwp/api/windows.security.authentication.web.webauthenticationbroker?view=winrt-22621)
API. We've opted to follow OAuth best practices more closely - e.g. using user's default browser as part of version 1. The best practices for the API are taken from the
IETF(Internet Engineering Task Force) OAuth 2.0 Authorization Framework [RFC 6749](https://tools.ietf.org/html/rfc6749),
Proof Key for Code Exchange (PKCE) [RFC 7636](https://tools.ietf.org/html/rfc7636),
and OAuth 2.0 for Native Apps [RFC 8252](https://tools.ietf.org/html/rfc8252).

It is recommended to read through the mentioned RFCs for better understanding of the design used in the API spec and the implementation code.
Please go through the API spec to understand the implementation.

**Testing**
Successful local build
Produced an experimental NuGet package leveraging the aggregator build pipeline, and tested it using a sample app.

Note: The Interactive Experiences package has been incorporated as a dependency in the Foundation repo to utilize Microsoft.UI API features.

**Pending Action Items:**
Ensure that Maestro recognizes the dependency updates being made to the Foundation branch (the Interactive Experiences package has been integrated into the Foundation repo)
Unit tests
Publishing Public/Private symbols
Create sample apps for WinAppSDK-samples repo.
Test support for C# projections

A microsoft employee must use /azp run to validate using the pipelines below.

WARNING:
Comments made by azure-pipelines bot maybe inaccurate.
Please see pipeline link to verify that the build is being ran.

For status checks on the main branch, please use TransportPackage-Foundation-PR
(https://microsoft.visualstudio.com/ProjectReunion/_build?definitionId=81063&_a=summary)
and run the build against your PR branch with the default parameters.

### CODE REVIEW

**Comment by akanpatel** [comment]
Samples: https://github.com/microsoft/WindowsAppSDK-Samples/tree/user/akanpatel2206/OAuth2_samples/Samples/OAuth
It will be updated to include more examples.

File: dev/OAuth/OAuth2Manager.cpp [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        auto paramsImpl = winrt::get_self<implementation::TokenRequestParams>(params);
+        paramsImpl->finalize();
+        if (clientAuth)
+        {
+            OAuth2ManagerTelemetry::RequestTokenAsyncTriggered(isAppPackaged, appName, paramsImpl->GrantType().c_str(), true);
+        }
+        else
+        {
+            OAuth2ManagerTelemetry::RequestTokenAsyncTriggered(isAppPackaged, appName, paramsImpl->GrantType().c_str(), false);
+        }
```
Comment (jonwis) [inline dev/OAuth/OAuth2Manager.cpp:276]:
```c++
OAuth2ManagerTelemetry::RequestTokenAsyncTriggered(isAppPackaged, appName, paramsImpl->GrantType().c_str(), clientAuth ? true : false);
```

File: dev/OAuth/OAuth2Manager.cpp [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                else
+                {
+                    auto success = winrt::make<implementation::TokenResponse>(jsonObject);
+                    co_return winrt::make<implementation::TokenRequestResult>(std::move(response), std::move(success),
+                        nullptr);
+                }
+            }
+            catch (...)
+            {
+                co_return implementation::TokenRequestResult::MakeFailure(std::move(response),
```
Comment (jonwis) [inline dev/OAuth/OAuth2Manager.cpp:368]: For these "catch and return" points, consider using `LOG_CAUGHT_EXCEPTION()` so you can see what's happening.

File: dev/OAuth/OAuth2Manager.cpp [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+            *itr = std::move(m_pendingAuthRequests.back());
+            m_pendingAuthRequests.pop_back();
+        }
+
+        return result;
+    }
+
+	std::wstring OAuth2Manager::create_implicit_url(const foundation::Uri& completeAuthEndpoint, const winrt::hstring& state, const foundation::Uri& redirectUri)
+	{
+		std::lock_guard guard{ m_mutex };
```
Comment (jonwis) [inline dev/OAuth/OAuth2Manager.cpp:437]: More indentation/whitespace to fix here

File: Directory.Packages.props [hunk 4]
```diff
@@ -6,6 +6,7 @@
 -->
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <ItemGroup>
+    <PackageVersion Include="Microsoft.ProjectReunion.InteractiveExperiences.TransportPackage" Version="1.8.0-CI-26107.1720.241003-1631.3" />
```
Comment (Saharsh979) [inline Directory.Packages.props:9]: Did this get automatically added when you added a reference to this dependency for the IXP packagereference?

File: dev/OAuth/OAuth2Manager.cpp (hunk 2 above)
Comment (Saharsh979) [reply dev/OAuth/OAuth2Manager.cpp:368]: The official Winappsdk docs mention how we can pull out the relevant hr code and message.
https://github.com/microsoft/WindowsAppSDK/blob/743f4551dcd8eda8db1136d7f61a8ecc7f338325/docs/Coding-Guidelines/Languages-CPP.md

Would it also be helpful if we can fire a telemetry with the error codes for debugging and tracking metrics?

File: Directory.Packages.props (hunk 4 above)
Comment (akanpatel2206) [reply Directory.Packages.props:9]: I made change to version.details.xml and synced dependencies.

File: dev/OAuth/OAuth2Manager.cpp (hunk 1 above)
Comment (akanpatel) [reply dev/OAuth/OAuth2Manager.cpp]: Done

File: dev/OAuth/OAuth2Manager.cpp (hunk 3 above)
Comment (akanpatel) [reply dev/OAuth/OAuth2Manager.cpp]: Done.

File: build/CopyFilesToStagingDir.ps1 [hunk 5]
```diff
@@ -186,6 +189,7 @@ PublishFile $FullBuildOutput\WindowsAppRuntime_DLL\StrippedWinMD\Microsoft.Windo
 PublishFile $FullBuildOutput\WindowsAppRuntime_DLL\StrippedWinMD\Microsoft.Windows.Management.Deployment.winmd $NugetDir\lib\uap10.0
 PublishFile $FullBuildOutput\WindowsAppRuntime_DLL\StrippedWinMD\Microsoft.Windows.PushNotifications.winmd $NugetDir\lib\uap10.0
 PublishFile $FullBuildOutput\WindowsAppRuntime_DLL\StrippedWinMD\Microsoft.Windows.Security.AccessControl.winmd $NugetDir\lib\uap10.0
+PublishFile $FullBuildOutput\WindowsAppRuntime_DLL\StrippedWinMD\Microsoft.Security.Authentication.OAuth.winmd $NugetDir\lib\uap10.0
```
[the remark this answers is not in the document]
Comment (AjitSurana) [reply build/CopyFilesToStagingDir.ps1:192]: Hello @jonwis, 
@akanpatel2206 put initially the name Microsoft.Windows.Security...  but I suggested to remove Windows out of it. The reason is these APIs don't have any dependency on Windows auth APIs underneath and the overall namespace name would be shorten. Do you think Keeping Microsoft.Security... namespace would be better than Microsoft.Windows.Security...?

File: dev/OAuth/AuthFailure.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                }
+                else
+                {
+                    additionalParams.emplace(std::move(name), entry.Value());
+                }
+            }
+        };
+
+        parseComponents(responseUri.Query());
+        parseComponents(fragment_component(responseUri));
```
Comment (AjitSurana) [inline dev/OAuth/AuthFailure.cpp:54]: Why you need to call the parseComponents two times?

File: dev/OAuth/AuthRequestAsyncOperation.cpp [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Security::Cryptography;
+
+AuthRequestAsyncOperation::AuthRequestAsyncOperation(const winrt::hstring& state)
+{
+    try
+    {
+        if (state.empty())
+        {
+            while (true)
```
Comment (AjitSurana) [inline dev/OAuth/AuthRequestAsyncOperation.cpp:22]: Why while loop is written here?

File: dev/OAuth/AuthRequestAsyncOperation.cpp [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+}
+
+void AuthRequestAsyncOperation::Cancel()
+{
+    winrt::make_self<factory_implementation::OAuth2Manager>()->cancel(this);
+}
+
+void AuthRequestAsyncOperation::Close()
+{
+    // TODO? C++/WinRT does a noop here
```
Comment (AjitSurana) [inline dev/OAuth/AuthRequestAsyncOperation.cpp:201]: What is this comment? Do we need the Close() function?

File: dev/OAuth/AuthFailure.cpp (hunk 6 above)
Comment (akanpatel) [reply dev/OAuth/AuthFailure.cpp:54]: If we have a URI like https://example.com/callback#access_token=abc123, calling fragment_component on this URI would return access_token=abc123.
https://www.rfc-editor.org/rfc/rfc6749#section-4.2.2
image

File: dev/OAuth/AuthRequestAsyncOperation.cpp (hunk 8 above)
Comment (akanpatel) [reply dev/OAuth/AuthRequestAsyncOperation.cpp:201]: The current implementation does nothing (a no-op, or "no operation") in this context.
This method is part of the IAsyncInfo interface.

File: dev/OAuth/AuthRequestAsyncOperation.cpp (hunk 7 above)
Comment (akanpatel) [reply dev/OAuth/AuthRequestAsyncOperation.cpp:22]: Updated the code to have state inside the loop. The loop is here to ensure we get a unique state value.

File: build/CopyFilesToStagingDir.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (akanpatel) [reply build/CopyFilesToStagingDir.ps1:192]: As discussed, keeping Microsoft.Security..

File: dev/OAuth/AuthRequestAsyncOperation.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        winrt::hresult hr = {});
+    void invoke_handler(const foundation::AsyncOperationCompletedHandler<oauth::AuthRequestResult>& handler);
+
+    void destroy();
+
+    std::shared_mutex m_mutex;
+
+    winrt::com_ptr<oauth::implementation::AuthRequestParams> m_params;
+    std::wstring m_pipeName;
+    HANDLE m_pipe = INVALID_HANDLE_VALUE;
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/AuthRequestAsyncOperation.h]: Made the change.

File: dev/OAuth/AuthRequestAsyncOperation.h [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    request_state m_state = request_state::connecting;
+    OVERLAPPED m_overlapped = {};
+    PTP_WAIT m_ptp = nullptr;
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/AuthRequestAsyncOperation.h]: Done.

File: dev/OAuth/OAuth2Manager.cpp [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+
+        try
+        {
+            // Pipe server has been successfully set up. Initiate the launch
+            auto url = paramsImpl->create_url(authEndpoint);
+
+            // Convert parentWindowId to HWND
+            HWND hwndParent = reinterpret_cast<HWND>(parentWindowId.Value);
+
+            auto launchResult = ::ShellExecuteW(hwndParent, L"open", url.c_str(), nullptr, nullptr, SW_SHOWDEFAULT);
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/OAuth2Manager.cpp]: Done.

File: dev/OAuth/OAuth2Manager.cpp [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+        // URL, these steps are intended more as a defense in depth. Other features such as PKCE should be used to
+        // ensure that codes/tokens are safe in the event that the state is compromised.
+        auto encryptedUri = encrypt(responseUri.RawUri(), state);
+
+        // When we create the named pipe, we only allow a single pipe instance. This should be fine under normal
+        // circumstances, however it might be the case that another process attaches to the pipe. This may be
+        // innocuous - e.g. the browser did multiple redirects - or it could be a bad actor - e.g. a process sending
+        // random garbage to any pipe it can open or another process specifically targeting oauth. Therefore we make
+        // multiple attempts to connect to the pipe
+        HANDLE pipe = INVALID_HANDLE_VALUE;
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/OAuth2Manager.cpp]: Done.

File: dev/OAuth/OAuth2Manager.cpp (hunk 2 above)
Comment (akanpatel2206) [reply dev/OAuth/OAuth2Manager.cpp:344]: Used LOG_CAUGHT_EXCEPTION();

File: dev/OAuth/AuthRequestParams.cpp [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    void AuthRequestParams::finalize()
+    {
+        std::lock_guard guard{ m_mutex };
+        if (m_finalized)
+        {
+            throw winrt::hresult_illegal_method_call(L"AuthRequestParams can only be used for a single request call");
+        }
+
+        m_finalized = true;
+        m_additionalParams->lock();
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/AuthRequestParams.cpp:164]: Documenting makes sense yes, but kept lock to be on the safe side.

File: dev/OAuth/OAuth2Manager.cpp [hunk 14]
```diff
@@ -0,0 +1,458 @@
...
+        const Uri& completeAuthEndpoint)
+    {
+        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Security::Authentication::OAuth::Feature_OAuth::IsEnabled());
+
+
+        bool isAppPackaged = m_telemetryHelper.IsPackagedApp();
+        PCWSTR appName = m_telemetryHelper.GetAppName().c_str();
+        OAuth2ManagerTelemetry::RequestAuthAsyncTriggered(isAppPackaged, appName, false);
+
+		winrt::hstring state{ random_base64urlencoded_string(32) };
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/OAuth2Manager.cpp]: Done.

File: dev/OAuth/OAuth2Manager.cpp [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+
+
+        bool isAppPackaged = m_telemetryHelper.IsPackagedApp();
+        PCWSTR appName = m_telemetryHelper.GetAppName().c_str();
+
+        auto paramsImpl = winrt::get_self<implementation::AuthRequestParams>(params);
+        auto asyncOp = winrt::make_self<AuthRequestAsyncOperation>(paramsImpl);
+		OAuth2ManagerTelemetry::RequestAuthWithParamsAsyncTriggered(isAppPackaged, appName, paramsImpl->ResponseType().c_str());
+
+
```
[the remark this answers is not in the document]
Comment (akanpatel2206) [reply dev/OAuth/OAuth2Manager.cpp:109]: Created another method for execute_shell and is called from all three functions.

File: dev/OAuth/AuthRequestParams.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (jonwis) [reply dev/OAuth/AuthRequestParams.cpp:164]: ... I think you should consider removing the lockable type; it adds complexity to the implementation. No change required, but for release have a hard look at whether the new complexity is adding value.

File: dev/OAuth/AuthRequestParams.cpp [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+        m_codeChallenge = std::move(value);
+    }
+
+    std::wstring AuthRequestParams::create_url(const Uri& authEndpoint)
+    {
+        std::shared_lock guard{ m_mutex };
+        WINRT_ASSERT(m_finalized);
+
+        // Per RFC 6749 section 3.1, the auth endpoint URI *MAY* contain a query string, which must be retained
+        std::wstring result{ authEndpoint.RawUri() };
```
Comment (jonwis) [inline dev/OAuth/AuthRequestParams.cpp:194]: I'm surprised there isn't a "URI builder" that would do this for you! Looks fine as-is.

**Review by jonwis — APPROVED** [review]
Thanks!

File: dev/Common/Common.vcxitems.filters [hunk 17]
```diff
@@ -43,6 +43,9 @@
     </ClInclude>
     <ClInclude Include="$(MSBuildThisFileDirectory)winrt_WindowsAppRuntime.h">
       <Filter>Header Files</Filter>
+    </ClInclude>
+	<ClInclude Include="$(MSBuildThisFileDirectory)TelemetryHelper.h">
```
Comment (Hemantxk) [inline dev/Common/Common.vcxitems.filters:47]: nit:space

File: dev/Common/Common.vcxitems [hunk 18]
```diff
@@ -27,6 +27,7 @@
     <ClInclude Include="$(MSBuildThisFileDirectory)Microsoft.Utf8.h" />
     <ClInclude Include="$(MSBuildThisFileDirectory)NotificationTelemetryHelper.h" />
     <ClInclude Include="$(MSBuildThisFileDirectory)Security.IntegrityLevel.h" />
+	<ClInclude Include="$(MSBuildThisFileDirectory)TelemetryHelper.h" />
```
Comment (Hemantxk) [inline dev/Common/Common.vcxitems:30]: nit:space

Comment (akanpatel2206) [reply dev/Common/Common.vcxitems]: Done.

File: dev/Common/Common.vcxitems.filters (hunk 17 above)
Comment (akanpatel2206) [reply dev/Common/Common.vcxitems.filters]: Done.

**Pull request merged**